cpp/src/sampling/random_walks_impl.cuh

```diff
@@ -84,7 +221,8 @@ struct uniform_selector {
     GraphViewType const& graph_view,
     std::optional<edge_property_view_t<typename GraphViewType::edge_type, weight_t const*>>
       edge_weight_view,
-    rmm::device_uvector<typename GraphViewType::vertex_type> const& current_vertices)
+    rmm::device_uvector<typename GraphViewType::vertex_type> const& current_vertices,
+    std::optional<rmm::device_uvector<typename GraphViewType::vertex_type>> &previous_vertices)
```

Comment:
Any reason to use `rmm::device_uvector const&` over `raft::device_span`?

Reply:
And for `previous_vertices`, use `std::optional<rmm::device_uvector<typename GraphViewType::vertex_type>> &&` instead of `std::optional<rmm::device_uvector<typename GraphViewType::vertex_type>> &`.  (&& instead of just &).

See https://stackoverflow.com/questions/37935393/pass-by-value-vs-pass-by-rvalue-reference

You're passing the ownership of `previous_vertices` to `folow_random_edge` so `follow_random_edge` can pass it pack to the caller.

Reply:
> Any reason to use rmm::device_uvector const& over raft::device_span?
For the uniform and biased selectors, there is not. But for the node2vec selector, on Line 518, the copy requires that current vertices be a `rmm::device_uvector` type.  Since each selector uses the same call, they all have the `rmm::device_uvector` type. If there is a way to avoid this, I am unaware of it but would love to learn.

Reply:
Interestingly, the uniform and biased selectors pass `rmm::device_uvector<vertex_t> const&`, where the node2vec selector passes `rmm::device_uvector<vertex_t>&` and has a side effect of doing `std::move` on it and invalidating the contents.  So it's not really the same.

I'd suggest marking this with a FIXME (in all 3 places).  My inclination would be to pass a span and do something different in the node2vec implementation (perhaps populate a new `rmm::device_uvector` or perhaps moving `rmm::device_uvector` in the function... we need to look at the context of what's going on to decide).